### PR TITLE
chore: unequate reservation existence and the fact they've been started

### PR DESCRIPTION
**Pull request by stoksc**
## Description
This change adds a new bool to the allocation actor to track when reservations have been started. `test_cancel_ten_experiment` was flaking occasionally, and it was caused by falsely equating `len(reservations) > 0` to "there are containers we should try to kill, and we should wait until we receive receipt of them being killed". This wasn't true when they are set at the top of `ResourcesAllocated` and we exited due to a DB failure, the experiment already being canceled, etc, before starting them. This caused allocations to hang waiting for `ContainerStopped` messages, which agents never send if you request a non-existent container to stop. This should make this happen less but still isn't perfect; in the event a container fails to launch this may still happen (since we would think we started it, but didn't). But, when starting reservations is truly an ask that can return all errors from the agent, this works (everything being completely async is slightly painful). This work is happening separately.

## Test Plan
- [x] run tests. since this is an inter-component bug that only happens in certain order of ops, very hard to write a test for, without forcing it to an extent it is unhelpful.

alternative solution is to always send back an `InvalidContainerRequest` (just sending a fake `ContainerStopped` is annoying because everywhere between the container and allocation actors try to read something off it) and something in the allocation actor handle it accordingly, but it'd be sweet if FT/reliability could _not_ be pushed up into lots of application layer specific logic.

relevant log snippet:
```
master: INFO[2022-02-04T17:39:19Z] experiment state changed to STOPPING_CANCELED  id=20 system=master type=experiment
master: INFO[2022-02-04T17:39:19Z] starting checkpoint garbage collection        id=experiment-18-checkpoint-gc system=master type=checkpointGCTask
master: DEBU[2022-02-04T17:39:19Z] patching state after request (STOPPING_CANCELED)  experiment-id=20 id=e01ed4fe-7a6a-4d89-ba95-46d0f9587713 system=master type=trial
master: INFO[2022-02-04T17:39:19Z] trial changed from state ACTIVE to STOPPING_CANCELED  experiment-id=20 id=e01ed4fe-7a6a-4d89-ba95-46d0f9587713 system=master type=trial
master: INFO[2022-02-04T17:39:19Z] decided to terminate trial                    experiment-id=20 id=e01ed4fe-7a6a-4d89-ba95-46d0f9587713 system=master type=trial
master: ERRO[2022-02-04T17:39:19Z] allocation encountered fatal error            error="could not get task spec: the allocation was canceled while this message was waiting" id=1 system=master type=Allocation
master: INFO[2022-02-04T17:39:19Z] decided to kill allocation                    id=1 system=master type=Allocation
master: DEBU[2022-02-04T17:39:19Z] still inside of kill cooldown                 id=1 system=master type=Allocation
master: INFO[2022-02-04T17:39:19Z] killing container id: 06fa4da6-e102-4f34-a0d7-8a8c6bb6d822  id=agent1 system=master type=agent
agent1: WARN[2022-02-04T17:39:19Z] error signaling container with killed, container not found: 06fa4da6-e102-4f34-a0d7-8a8c6bb6d822  id=containers system=agent1 type=containerManager
```

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by stoksc** [comment]
open to discuss alternatives here, lmk what you think @ioga

File: master/internal/task/allocation.go [hunk 1]
```diff
@@ -353,6 +355,7 @@ func (a *Allocation) ResourcesAllocated(ctx *actor.Context, msg sproto.Resources
 			IsMultiAgent: len(a.reservations) > 1,
 		})
 	}
+	a.reservationsStarted = true
```
Comment (ioga) [inline master/internal/task/allocation.go:358]: Let's say
1. in line 352 here,`r.Start`, tells agent actor to start a task
2. we set reservationStarted here.
3. agent network connection breaks and it cannot write message # 1

then I suspect in the current code we'll *probably* receive `ContainerStopped` when agent actor will die one way or another?... is that right?

**Review by ioga — APPROVED** [review]

File: master/internal/task/allocation.go (hunk 1 above)
Comment (stoksc) [reply master/internal/task/allocation.go:358]: yeah, that is basically my thought exactly

Comment (stoksc) [reply master/internal/task/allocation.go:358]: so long as we dont fail without handling it correctly, other downstream failures should cause us to handle this correctly

**Pull request merged**